### PR TITLE
feat: zero-sized queue

### PR DESCRIPTION
**Pull request by 3Hren**
This commit introduces zero-sized queue for App service. It allows to
immediately inject events without (almost) putting them into the queue.
Also if all slaves are busy the new policy will reject new events
without putting them into the queue.

### CODE REVIEW

File: node/src/node/engine.cpp [hunk 1]
```diff
@@ -178,27 +179,50 @@ auto engine_t::enqueue(event_t event, std::shared_ptr<api::stream_t> rx)
...
+                    std::move(event),
+                    trace_t::current(),
+                    tx->dispatch, // Explicitly copy.
+                    std::move(rx)
+                };
 
-            tx->dispatch = std::make_shared<client_rpc_dispatch_t>(manifest().name);
+                if (limit > 0) {
+                    queue.push_back(load);
+                    stats.queue_depth->add(queue.size());
```
Comment (antmat) [inline node/src/node/engine.cpp]: Looks like it can be moved down

File: node/src/node/engine.cpp [hunk 2]
```diff
@@ -395,90 +416,116 @@ auto engine_t::on_spawn_rate_timeout(const std::error_code&) -> void {
...
-    using boost::adaptors::filtered;
-
-    const auto concurrency = profile().concurrency;
-
-    // Find an active non-overloaded slave.
-    const auto filter = std::function<bool(const slave_t&)>([&](const slave_t& slave) -> bool {
+auto engine_t::select_slave(pool_type& pool) -> boost::optional<slave_t&> {
+    auto concurrency = profile().concurrency;
+    return select_slave(pool, [&](const slave_t& slave) -> bool {
         return slave.active() && slave.load() < concurrency;
```
Comment (antmat) [inline node/src/node/engine.cpp:413]: BTW, why not least concurency of spawned slaves?

File: node/src/node/engine.cpp [hunk 3]
```diff
@@ -490,24 +537,20 @@ auto engine_t::rebalance_slaves() -> void {
 
                 while (active-- > target) {
                     // Find active slave with minimal load.
```
Comment (antmat) [inline node/src/node/engine.cpp:533]: comment says about minimal load. It's not true, is it?

File: node/src/node/engine.cpp [hunk 4]
```diff
@@ -395,90 +416,116 @@ auto engine_t::on_spawn_rate_timeout(const std::error_code&) -> void {
...
+
+    std::size_t target;
+    if (manual_target > 0) {
+        target = manual_target;
+    } else {
+        if (profile.queue_limit > 0) {
+            target = load / profile.grow_threshold;
+        } else {
+            target = pool.apply([&](pool_type& pool) {
+                auto pressure = boost::accumulate(pool |
```
Comment (antmat) [inline node/src/node/engine.cpp]: looks like vacant calculation code duplication

**Review by antmat — COMMENTED** [review]
Looks ok, except minor notices

**Comment by 3Hren** [comment]
Fixed!

**Review by antmat — APPROVED** [review]

**Pull request merged**